src/audio/codec_adapter/codec_adapter.c

```diff
@@ -0,0 +1,515 @@
+				sink->w_ptr));
+	uint32_t tail_size = bytes - head_size;
+
+	memcpy(sink->w_ptr, source, head_size);
+	if (tail_size)
+		memcpy(audio_stream_wrap(sink, (char *)sink->w_ptr + head_size),
+		       (char *)source + head_size,
+		       tail_size);
+}
+
+static int codec_adapter_copy(struct comp_dev *dev)
+{
+	int ret = 0;
+	uint32_t bytes_to_process, processed = 0;
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct codec_data *codec = &cd->codec;
+	struct comp_buffer *source = cd->ca_source;
+	struct comp_buffer *sink = cd->ca_sink;
+	uint32_t codec_buff_size = codec->cpd.in_buff_size;
+	struct comp_copy_limits cl;
+
+	comp_get_copy_limits_with_lock(source, sink, &cl);
+	bytes_to_process = cl.frames * cl.source_frame_bytes;
+
+	comp_dbg(dev, "codec_adapter_copy() start: codec_buff_size: %d, sink free: %d source avail %d",
+		 codec_buff_size, sink->stream.free, source->stream.avail);
+
+	while (bytes_to_process) {
+		/* Proceed only if we have enough data to fill the lib buffer
+		 * completely. If you don't fill whole buffer
+		 * the lib won't process it.
+		 */
+		if (bytes_to_process < codec_buff_size) {
+			comp_dbg(dev, "codec_adapter_copy(): processed %d in this call %d bytes left for next period",
+				 processed, bytes_to_process);
+			break;
+		}
+
+		codec_adapter_copy_to_lib(&source->stream,
+					  codec->cpd.in_buff,
+					  codec_buff_size);
+		buffer_invalidate(source, codec_buff_size);
+		codec->cpd.avail = codec_buff_size;
+		//TODO: call codec specific processing here
+		codec_adapter_copy_from_lib_to_sink(codec->cpd.out_buff,
+						    &sink->stream,
+						    codec->cpd.produced);
+		buffer_writeback(sink, codec->cpd.produced);
+
+		bytes_to_process -= codec->cpd.produced;
+		processed += codec->cpd.produced;
+	}
+
+	if (!processed) {
+		comp_dbg(dev, "codec_adapter_copy(): nothing processed in this call");
+		goto end;
+	}
+
+	comp_update_buffer_produce(sink, processed);
+	comp_update_buffer_consume(source, processed);
+end:
+	return ret;
+}
+
+static int codec_adapter_set_params(struct comp_dev *dev, struct sof_ipc_ctrl_data *cdata,
+			 enum codec_cfg_type type)
+{
+	int ret;
+	char *dst, *src;
+	static uint32_t size;
+	uint32_t offset;
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	comp_dbg(dev, "codec_adapter_set_params(): start: num_of_elem %d, elem remain %d msg_index %u",
+		 cdata->num_elems, cdata->elems_remaining, cdata->msg_index);
+
+	/* Stage 1 verify input params & allocate memory for the config blob */
+	if (cdata->msg_index == 0) {
+		size = cdata->num_elems + cdata->elems_remaining;
+		/* Check that there is no work-in-progress on previous request */
+		if (cd->runtime_params) {
+			comp_err(dev, "codec_adapter_set_params() error: busy with previous request");
+			ret = -EBUSY;
+			goto end;
+		} else if (!size) {
+			comp_err(dev, "codec_adapter_set_params() error: no configuration size %d",
+				 size);
+			//TODO: return -EINVAL. This is temporary until driver fixes its issue
```

Comment:
Do we have a driver bug reported for this ? @kv2019i @juimonen fyi.

Reply:
Not yet, but I am preparing a patch for it.

Reply:
any update ?